sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala

```diff
@@ -0,0 +1,497 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids.shims
+
+import ai.rapids.cudf.{BinaryOperable, ColumnVector, DType, RoundMode, Scalar}
+import com.nvidia.spark.rapids.{Arm, GpuBinaryExpression, GpuColumnVector, GpuScalar}
+
+import org.apache.spark.sql.catalyst.expressions.{Expression, ImplicitCastInputTypes, NullIntolerant}
+import org.apache.spark.sql.types._
+
+object IntervalUtils extends Arm {
+
+  /**
+   * compute interval.toLong * num
+   */
+  def mulIntervalAndNum(interval: BinaryOperable, num: BinaryOperable): ColumnVector = {
+    interval match {
+      case intervalVector: ColumnVector =>
+        withResource(intervalVector.castTo(DType.INT64)) { intervalLong =>
+          intervalLong.mul(num)
+        }
+      case intervalScalar: Scalar =>
+        withResource(Scalar.fromLong(getLong(intervalScalar))) { intervalScalar =>
+          intervalScalar.mul(num)
+        }
+    }
+  }
+
+  /**
+   * throws exception if any value in `longCv` exceeds the int limits
+   */
+  def toIntWithCheck(longCv: ColumnVector): ColumnVector = {
+    withResource(longCv.castTo(DType.INT32)) { intResult =>
+      withResource(longCv.notEqualTo(intResult)) { notEquals =>
+        if (notEquals.any().getBoolean) {
+          throw new ArithmeticException("overflow occurs")
+        } else {
+          intResult.incRefCount()
+        }
+      }
+    }
+  }
+
+  /**
+   * Check if the long result of long * long is overflow
+   * See Math.multiplyExact:
+   * r = x * y;
+   * if (y != 0) && (r/y != x) throw exception;
+   * if(x == Long.MIN_VALUE && y == -1) throw exception
+   *
+   * @param leftLong  long cv or scalar
+   * @param rightLong long cv or scalar, will not be scalar if leftLong is scalar
+   * @param result    is the result of x * y
+   */
+  def checkMultiplyOverflow(leftLong: BinaryOperable, rightLong: BinaryOperable,
+      result: ColumnVector): Unit = {
+    leftLong match {
+      case leftVector: ColumnVector =>
+        checkMultiplyOverflowImp(rightLong, leftVector, result)
+      case _ =>
+        checkMultiplyOverflowImp(leftLong, rightLong.asInstanceOf[ColumnVector], result)
+    }
+  }
+
+  /**
+   * Check if the long result of long * long is overflow
+   * See Math.multiplyExact:
+   * r = x * y;
+   * if (y != 0) && (r/y != x) throw exception;
+   * if(x == Long.MIN_VALUE && y == -1) throw exception
+   *
+   * @param x long cv or scalar
+   * @param y long cv or scalar, will not be scalar if x is scalar
+   * @param r is the result of x * y
+   */
+  def checkMultiplyOverflowImp(x: BinaryOperable, y: ColumnVector,
```

Comment:
Why do we have so many functions for the overflow check ? Could we merge them into single one ?

Reply:
There are four types of checking,  seems can not merge to one.

**`day-time interval` * num**
The internal type of `day-time interval` is Long
The check of `r = x(long day-time  value) * y( (byte, short, int or long) value)` is:

```
   * r = x * y;
   * if (y != 0) && (r/y != x) throw exception;
   * if(x == Long.MIN_VALUE && y == -1) throw exception   
```

The check of r = x(long day-time  value) * y(double value) is:
```
See: com.google.common.math.roundToLong
   * r = x * y;
   * has no Inf, -Inf, NaN
   * z = roundIntermediate(x, mode)
   * checkInRange(MIN_LONG_AS_DOUBLE - z < 1.0 & z < MAX_LONG_AS_DOUBLE_PLUS_ONE)
   * return z.toLong
```



**`year-month interval` * num**
The internal type of `year-month interval` is int
The check of `r = x(int year-month value) * y( (byte, short, int or long) value)` is:

```
   * r = x * y;
   * if (y != 0) && (r/y != x) throw exception;
   * check (int) r == r 
```
Note:
~~if(x == Long.MIN_VALUE && y == -1) throw exception~~
No need to check the `if(x == Long.MIN_VALUE && y == -1) throw exception`, 
replaced by checking `(int) r == r`, because of expecting int value.
If (x,y) is (Long.MIN_VALUE, -1) or (-1, Long.MIN_VALUE) `check (int) r == r` will cover

The check of r = x(double value) * y(double value) is:
```
   * r = x * y;
   * has no Inf, -Inf, NaN
   * z = roundIntermediate(x, mode)
   * checkInRange(MIN_LONG_AS_DOUBLE - z < 1.0 & z < MAX_LONG_AS_DOUBLE_PLUS_ONE)
   * return z.toInt.    // Note this is toInt
```